realm/realm-library/src/main/java/io/realm/Realm.java

```diff
@@ -1366,23 +1365,30 @@ public RealmAsyncTask executeTransactionAsync(final Transaction transaction, fin
      * @throws IllegalArgumentException if the {@code transaction} is {@code null}, or if the realm is opened from
      *                                  another thread.
      */
-    public RealmAsyncTask executeTransactionAsync(final Transaction transaction, final Realm.Transaction.OnSuccess onSuccess, final Realm.Transaction.OnError onError) {
+    public RealmAsyncTask executeTransactionAsync(final Transaction transaction,
+                                                  final Realm.Transaction.OnSuccess onSuccess,
+                                                  final Realm.Transaction.OnError onError) {
         checkIfValid();
 
         if (transaction == null) {
             throw new IllegalArgumentException("Transaction should not be null");
         }
 
-        // If the user provided a Callback then we make sure, the current Realm has a Handler
-        // we can use to deliver the result
-        if ((onSuccess != null || onError != null) && !hasValidNotifier()) {
-            throw new IllegalStateException("Your Realm is opened from a thread without a Looper" +
-                    " and you provided a callback, we need a Handler to invoke your callback");
+        // Avoid to call canDeliverNotification() in bg thread.
+        final boolean canDeliverNotification = sharedRealm.capabilities.canDeliverNotification();
+
+        // If the user provided a Callback then we have to make sure the current Realm has an events looper to deliver
+        // the results.
+        if ((onSuccess != null || onError != null)  && !canDeliverNotification) {
+            throw new IllegalStateException("Your Realm is opened from a thread without an event looper." +
```

Comment:
"from a non-looper thread" would make more sense for Android developers I think.